### ISSUE TITLE
Dunder __exit__ exception propagation not respected when __exit__ return type is explicitly set to Literal[False] - causes "must return value" errors.

### ISSUE THREAD

**Issue by laundmo** [microsoft/pyright#3665]
**Describe the bug**
When the `__exit__` of a context manager explicitly hints the return type of `__exit__` as `Literal[False]`, Pyright fails to assume all errors propagate outside the context manager. Without the explicit hint, it is treated correctly, and Pyright assumes all errors propagate.

Faulty report **with** explicit `Literal[False]`:
```
pyright 1.1.257
test.py
  test.py:17:13 - information: Type of "A.__exit__" is "(self: A, exc_type: Type[BaseException] | None, exc_value: BaseException | None, traceback: TracebackType | None) -> Literal[False]"
  test.py:19:12 - error: Function with declared type of "int" must return value
    Type "None" cannot be assigned to type "int" (reportGeneralTypeIssues)
```

Correct report **without** explicit `Literal[False]`:

```
pyright 1.1.257
test.py
  test.py:17:13 - information: Type of "A.__exit__" is "(self: A, exc_type: Type[BaseException] | None, exc_value: BaseException | None, traceback: TracebackType | None) -> Literal[False]"
```

**To Reproduce**
1. run pyright on the example code provided below.
2. remove the explicit ` -> Literal[False]` type hint of `__exit__`
3. run pyright again
4. observe the difference in errors even though the signature and return type of the function stayed the same.

**Expected behavior**

Pyright should not observe any difference between explicit and inferred types. Especially not one where the behavior for inferred types is correct and for explicit isn't.

**Screenshots or Code**

```python
from types import TracebackType
from typing import Literal, Optional, Type
import random

class A:
    def __enter__(self):
        pass

    def __exit__(
        self,
        exc_type: Optional[Type[BaseException]],
        exc_value: Optional[BaseException],
        traceback: Optional[TracebackType],
    ) -> Literal[False]:
        return False

reveal_type(A.__exit__)

def f() -> int:
    with A():
        if random.randint(0, 1) == 1:
            raise Exception("exception")
        return 1
```

**VS Code extension or command-line**
Error occurs in both:
pyright 1.1.257 CLI
pylance v2022.6.30

**Additional context**
[PEP343](https://peps.python.org/pep-0343/#specification-the-with-statement) on the return value of __exit__

**Comment by laundmo** [comment]
Nice, always love how fast issues here are resolved!

i wasn't aware of the standard, so i instinctively went off the vague knowledge that a false return means no swallow.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.258, which I just published. It will also be included in the next release of pylance.
